pkg/proxy/iptables/proxier.go

```diff
@@ -396,66 +411,56 @@ func CleanupLeftovers(ipt utiliptables.Interface) (encounteredError bool) {
-			if _, found := existingNATChains[chain]; found {
-				chainString := string(chain)
-				writeBytesLine(natChains, existingNATChains[chain]) // flush
-				writeLine(natRules, "-X", chainString)              // delete
+		for _, chain := range chains {
+			if err := ipt.FlushChain(utiliptables.TableNAT, chain); err != nil {
+				if !utiliptables.IsNotFoundError(err) {
+					klog.Errorf("Error flushing iptables chain %s: %v", chain, err)
+					encounteredError = true
+				}
 			}
 		}
-		// Hunt for service and endpoint chains.
-		for chain := range existingNATChains {
-			chainString := string(chain)
-			if strings.HasPrefix(chainString, "KUBE-SVC-") || strings.HasPrefix(chainString, "KUBE-SEP-") || strings.HasPrefix(chainString, "KUBE-FW-") || strings.HasPrefix(chainString, "KUBE-XLB-") {
-				writeBytesLine(natChains, existingNATChains[chain]) // flush
-				writeLine(natRules, "-X", chainString)              // delete
+	}
+	// Flush all of our "-t filter" chains.
+	for _, chain := range []utiliptables.Chain{kubeServicesChain, kubeExternalServicesChain, kubeForwardChain} {
+		if err := ipt.FlushChain(utiliptables.TableFilter, chain); err != nil {
+			if !utiliptables.IsNotFoundError(err) {
+				klog.Errorf("Error flushing iptables chain %s: %v", chain, err)
+				encounteredError = true
 			}
 		}
-		writeLine(natRules, "COMMIT")
-		natLines := append(natChains.Bytes(), natRules.Bytes()...)
-		// Write it.
-		err = ipt.Restore(utiliptables.TableNAT, natLines, utiliptables.NoFlushTables, utiliptables.RestoreCounters)
-		if err != nil {
-			klog.Errorf("Failed to execute iptables-restore for %s: %v", utiliptables.TableNAT, err)
-			encounteredError = true
-		}
 	}
+	return encounteredError
+}
 
-	// Flush and remove all of our "-t filter" chains.
-	iptablesData.Reset()
-	if err := ipt.SaveInto(utiliptables.TableFilter, iptablesData); err != nil {
-		klog.Errorf("Failed to execute iptables-save for %s: %v", utiliptables.TableFilter, err)
+// CleanupLeftovers removes all iptables chains created by the Proxier
+// It returns true if an error was encountered. Errors are logged.
+func CleanupLeftovers(ipt utiliptables.Interface) (encounteredError bool) {
+	// Delete all of our "-t nat" chains.
+	chains, err := getExistingKubeNATChains(ipt)
+	if err != nil {
 		encounteredError = true
 	} else {
-		existingFilterChains := utiliptables.GetChainLines(utiliptables.TableFilter, iptablesData.Bytes())
-		filterChains := bytes.NewBuffer(nil)
-		filterRules := bytes.NewBuffer(nil)
-		writeLine(filterChains, "*filter")
-		for _, chain := range []utiliptables.Chain{kubeServicesChain, kubeExternalServicesChain, kubeForwardChain} {
-			if _, found := existingFilterChains[chain]; found {
-				chainString := string(chain)
-				writeBytesLine(filterChains, existingFilterChains[chain])
-				writeLine(filterRules, "-X", chainString)
+		for _, chain := range chains {
+			if err := ipt.FlushChain(utiliptables.TableNAT, chain); err != nil {
```

Comment:
Looks like we're calling this twice now. Can we rename `CleanupLeftovers` to `DeleteIptablesChains` and have it only do delete?

Reply:
ok